### PR TITLE
AP_Baro: make backends responsible for setting their health

### PR DESCRIPTION
**Pull request by peterbarker**
This resolves a problem in SITL where the barometer would
not change value for prolonged period of time, making it
go unhealthy

### CODE REVIEW

File: libraries/AP_Baro/AP_Baro.cpp [hunk 1]
```diff
@@ -512,17 +512,10 @@ void AP_Baro::update(void)
 
     if (!_hil_mode) {
         for (uint8_t i=0; i<_num_drivers; i++) {
-            drivers[i]->update();
+            drivers[i]->backend_update(i);
```
Comment (magicrub) [inline libraries/AP_Baro/AP_Baro.cpp]: I like this new term "backend_update". We should use that terminology in all of the libraries!

Comment (peterbarker) [inline libraries/AP_Baro/AP_Baro.cpp]: Might be better to have `backends[i]->update()`, 'though! :-)

**Comment by khancyr** [comment]
seems good!

**Comment by lvale** [comment]
Thanks @peterbarker

**Comment by rmackay9** [comment]
Perhaps I'm slightly overstepping my bounds by merging this but did it anyway..

**Comment by OXINARF** [comment]
I hadn't reviewed this yet because I thought there might be some implications to it that I needed to look carefully. I've looked at it now and it looks like at least the SIM_BARO_DISABLE parameter will disable updates to values, but will now not make the baro unhealthy.

This changes the place where the healthy check is done, but it didn't change the actual problem with the SITL backend, it can be more than 2 seconds without changing pressure and temperature.

**Comment by peterbarker** [comment]
On Fri, 14 Jul 2017, Francisco Ferreira wrote:

> I hadn't reviewed this yet because I thought there might be some
> implications to it that I needed to look carefully. I've looked at it now
> and it looks like at least the SIM_BARO_DISABLE parameter will disable
> updates to values, but will now not make the baro unhealthy.

Shouldn't the BARO simply not be used in that case, so it doesn't matter 
if it is healthy/unhealthy?

> This changes the place where the healthy check is done, but it didn't change
> the actual problem with the SITL backend, it can be more than 2 seconds
> without changing pressure and temperature.

I think it has to stay that way, too - otherwise SITL becomes 
non-deterministic, which would be bad.  Otherwise I'd have just changed 
the noise parameter and been done with it ;-)

**Comment by OXINARF** [comment]
> Shouldn't the BARO simply not be used in that case, so it doesn't matter
if it is healthy/unhealthy?

When you say it like that, it looks like you are saying it should act like *baro doesn't exist* as opposed to *baro exists but isn't sending any information*. If this is what you are saying, then that can't happen for a simple reason: we assume a baro exists, always; otherwise we panic on initialization.

> I think it has to stay that way, too - otherwise SITL becomes
non-deterministic, which would be bad.  Otherwise I'd have just changed
the noise parameter and been done with it ;-)

I'm torn here. Although I agree that having it deterministic is the correct way, the truth is that a real sensor has some amount of noise, so SITL is now an imperfect simulation. SITL INS driver has a [minimum amount of noise](https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_InertialSensor/AP_InertialSensor_SITL.cpp#L54-L56) so I think it would be acceptable to have the same thing in baro.

**Comment by OXINARF** [comment]
Besides the problem identified above, this also added a new build warning, given that `instance` is shadowed in the SITL backend.

**Comment by rmackay9** [comment]
sorry!

**Pull request merged**